src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Win32.cs

```diff
@@ -245,95 +248,108 @@ private void ReadDirectoryChangesCallback(uint errorCode, uint numBytes, AsyncRe
                             oldName = ReadOnlySpan<char>.Empty;
-                            break;
-                        default:
-                            if (!oldName.IsEmpty)
-                            {
-                                // Previous FILE_ACTION_RENAMED_OLD_NAME with no new name
-                                NotifyRenameEventArgs(WatcherChangeTypes.Renamed, ReadOnlySpan<char>.Empty, oldName);
-                                oldName = ReadOnlySpan<char>.Empty;
-                            }
-
-                            switch (info->Action)
-                            {
-                                case Interop.Kernel32.FileAction.FILE_ACTION_ADDED:
-                                    NotifyFileSystemEventArgs(WatcherChangeTypes.Created, info->FileName);
-                                    break;
-                                case Interop.Kernel32.FileAction.FILE_ACTION_REMOVED:
-                                    NotifyFileSystemEventArgs(WatcherChangeTypes.Deleted, info->FileName);
-                                    break;
-                                case Interop.Kernel32.FileAction.FILE_ACTION_MODIFIED:
-                                    NotifyFileSystemEventArgs(WatcherChangeTypes.Changed, info->FileName);
-                                    break;
-                                default:
-                                    Debug.Fail($"Unknown FileSystemEvent action type!  Value: {info->Action}");
-                                    break;
-                            }
-
-                            break;
-                    }
+                        }
 
-                    info = info->NextEntryOffset == 0 ? null : (Interop.Kernel32.FILE_NOTIFY_INFORMATION*)((byte*)info + info->NextEntryOffset);
-                } while (info != null);
+                        switch (info.Action)
+                        {
+                            case Interop.Kernel32.FileAction.FILE_ACTION_ADDED:
+                                NotifyFileSystemEventArgs(WatcherChangeTypes.Created, fileName);
+                                break;
+                            case Interop.Kernel32.FileAction.FILE_ACTION_REMOVED:
+                                NotifyFileSystemEventArgs(WatcherChangeTypes.Deleted, fileName);
+                                break;
+                            case Interop.Kernel32.FileAction.FILE_ACTION_MODIFIED:
+                                NotifyFileSystemEventArgs(WatcherChangeTypes.Changed, fileName);
+                                break;
+                            default:
+                                Debug.Fail($"Unknown FileSystemEvent action type!  Value: {info.Action}");
+                                break;
+                        }
 
-                if (!oldName.IsEmpty)
+                        break;
+                }
+
+                if (info.NextEntryOffset == 0 || info.NextEntryOffset > (uint)buffer.Length)
```

Comment:
This fixes another potential problem where things can wrong in the original code. For example, the original code can get into infinite loop with corrupted data here. We will just exit the loop now.